### ISSUE TITLE
One-off tasks against a dependent container

### ISSUE THREAD

**Issue by jagregory** [charleskorn/batect#105]
I have a service and a local S3 instance. I need to run an aws-cli command to create a bucket before my app can run. What I would like to do is:

- Launch the local S3
- Run aws-cli command to create a bucket
- Launch my local server

How I’ve attempted it:

```yaml
containers:
  s3: ...
  aws-cli: ...
  service: ...

tasks:
  start:
    container: service
    dependencies:
      - s3
    prerequisites:
      - createS3Bucket
   
  createS3Bucket:
    container: aws-cli
    run:
      command: aws s3 mb s3://example
```

This doesn't work because the prerequisites are executed before the task's dependencies are started.

Then I tried adding the S3 container as a dependency to the `createS3Bucket` task:

```yaml
containers:
  s3: ...
  aws-cli: ...
  service: ...

tasks:
  start:
    container: service
    dependencies:
      - s3
    prerequisites:
      - createS3Bucket
   
  createS3Bucket:
    container: aws-cli
    dependencies:
      - s3
    run:
      command: aws s3 mb s3://example
```

This also doesn't work, because when `createS3Bucket` is complete it tears down the `s3` container, and then `start` spins up a new one.

It'd be great if Batect had a way to model this flow.

**Comment by charleskorn** [comment]
Hey @jagregory, thanks for the suggestion. I agree that it would be great if batect could cover this situation.

I'm going to think about this a bit more before I commit to a solution. One idea I have is something like a `setup_commands` array on each container definition, so then it would look something like this:

```yaml
containers:
  s3:
    ...
    setup_commands:
      - aws s3 mb s3://example

  service: ...

tasks:
  start:
    container: service
    dependencies:
      - s3
```

The idea being that these commands would be run in the containers after they report as healthy but before any containers that depend on them start. Would something like that fit your use case? 

Or would you rather have the ability to run a command in another container as a setup step? I'm imagining something like init containers in Kubernetes. That would be significantly more complex but also much more powerful and flexible:

```yaml

tasks:
  start:
    container: service
    dependencies:
      - s3
    setup:
      - container: aws-cli
        command: aws s3 mb s3://example
```

In the meantime, to solve your problem now, you might be able to create the bucket at image build time and bake it into the image if your S3 mock saves things to disk. Or you could do something a bit hacky where you have a intermediary init container that depends on the S3 container and creates the bucket and then sleeps and reports as healthy, which the service container then depends on - the effect being that the service won't start until the initialisation is complete.

**Comment by androa** [comment]
I've run into a similar case. I'm using http://www.mock-server.com/ to mock away some old and horrible SOAP-services while doing local development, and now it's bundled in together with the application. Which has it's pros and cons, but now I needed the same mocks in another application.

MockServer works by spinning up as server with an HTTP API to create new mocks and expectations. It holds no state for longer than the lifetime of the server.

The idea was to pull the scaffolding of the mocks out into a separate repo and using batect to spin up mockserver, run a one-off job to set up all the mocked services, and then run my application.

But now mockserver gets restarted between the task of setting up mocks, and the task for running the application, and therefore looses all the mocks.

**Comment by charleskorn** [comment]
Hey @androa, thanks for your feedback. Would the first solution in my comment above (`setup_commands`) be sufficient for you, or would you need the ability to spin up a second container to initialise MockServer?

**Comment by androa** [comment]
Yes and no. MockServer is a docker container with JDK available. So I can get around the initialisation by doing either:  

1. Extend the MockServer docker image and insert a whole Java project and run it. It's cumbersome to have to maintain a separate docker image just for this.
2. Mount the Java project/jar as a volume to the MockServer image and use the `setup_commands`.

I do however feel it's a lot of overhead to maintain the setup as Java project. Going with JavaScript is so significantly more lightweight. But that requires a node runtime, which in turn requires extension of the MockServer image to set up node.

Ideally it would be better to spin up an off the shelf node image, mount the code as a volume and just run the necessary setup.

After commenting in this issue, I've started thinking this might be such an edge case it's not really worth supporting. We currently have a bit of sunk cost into MockServer, but I'm leaning towards finding a different mocking solution like f.ex.`nock` or ´mockttp`. It seems to be fairly simple conversion :)

**Comment by charleskorn** [comment]
I've had a lot of success with Mountebank as well - there's an example of setting it up in the Java sample project (https://github.com/charleskorn/batect-sample-java/tree/master/.batect/exchange-rate-service-fake).

**Comment by priorax** [comment]
This would be really helpful to me as well.
We are currently in the middle of setting up our dotnet core projects to run against temporary Postgres databases.
Since we are wanting to test with fresh databases as well, we want to perform EntityFramework migrations after the dependent database has launched, but before we commence our integration tests.
At the moment we have tried a few variations, but can't find a nice way of handling it since the API server we are starting for doing our tests against is using the same Dockerfile we run in our CI to ship.

At the moment we have the following containers:

- postgres
- api
  - dependent on postgres
- test suite
  - dependent on api

We have tried a few things using prerequisites and dependencies, but keep coming back to the same issues mentioned above.

It would be nice to be able to run additional commands against the same postgres database _somehow_ (preferably in a new container that has the dotnet sdk, but I'm sure we can have some process work).

**Comment by charleskorn** [comment]
OK, it looks like there are two solutions here, both of which would be valuable:

* some way to run commands in a container that has just become healthy but before any containers that depend on it start (`setup_commands` from above - name TBC)
* some way to run commands in another container after another container becomes healthy but before any containers that depend on it start

The first of these is relatively straightforward to implement, so I'm planning on adding it soon.

The second is far more complex and has a lot of scenarios that make things more tricky (eg. what happens if that other container itself has dependencies), so that might be a longer-term feature.

**Comment by thelukemccarthy** [comment]
I have a similar problem. What I would like to be able to do is run the first task to completion then use batect to spin up another container that joins the same network.

Maybe a new field called 'network-group' could be used to specify which containers get added to the same network?

**Comment by charleskorn** [comment]
Hey @thelukemccarthy, could you expand a little on what you're trying to do? What does the first task do, and what does the second container do?

**Comment by charleskorn** [comment]
An update on this: I've added a basic version of this (the `setup_commands` approach discussed in [this comment](https://github.com/charleskorn/batect/issues/105#issuecomment-495142763)), and it's just been released as part of [v0.38](https://github.com/charleskorn/batect/releases/tag/0.38.0). 

I don't have any immediate plans to implement the more advanced scenario of running a setup command in another container, but I'll add it to the list of feature ideas. If anyone is keen to see this sooner, I'm always open to PRs :)

**Comment by charleskorn** [comment]
I'm going to close this issue in favour of continuing the discussion on the advanced scenario in #286.
